### PR TITLE
Cleanup metadata and add documentations for it

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR:

- adds documentation at `Metadata` with some doctests.
- renames:
  - `all_fields` -> `columns`
  - `index_fields` -> `index_columns`
  - `column_fields` -> `data_columns`
  - `index_names` -> `index_names` (no change)
  - `index_info` -> `index_map`

  I tried to keep the original intention and avoid name conflict between field and column.
Also, `index_info` -> `index_map`, for now (see https://github.com/databricks/koalas/pull/290#issuecomment-491450583)

- fixes a bug (one line fix) at `set_index`:

    ```python
    kdf = ks.DataFrame({
        'A': [1, 2, 3, 4],
        'B': [5, 6, 7, 8],
        'C': [9, 10, 11, 12],
        'D': [13, 14, 15, 16],
        'E': [17, 18, 19, 20]})
    kdf1 = kdf.set_index("A")
    kdf1.to_spark().show()
    ```

   ```
    Failed example:
        kdf1.to_spark().show()  # doctest: +NORMALIZE_WHITESPACE
    Expected:
        +---+---+---+---+---+
        |  A|  B|  C|  D|  E|
        +---+---+---+---+---+
        |  1|  5|  9| 13| 17|
        |  2|  6| 10| 14| 18|
        |  3|  7| 11| 15| 19|
        |  4|  8| 12| 16| 20|
        +---+---+---+---+---+
    Got:
        +-----------------+---+---+---+---+---+
        |__index_level_0__|  A|  B|  C|  D|  E|
        +-----------------+---+---+---+---+---+
        |                0|  1|  5|  9| 13| 17|
        |                1|  2|  6| 10| 14| 18|
        |                2|  3|  7| 11| 15| 19|
        |                3|  4|  8| 12| 16| 20|
        +-----------------+---+---+---+---+---+
   ```

    Note that `__index_level_0__` doesn't exist in metadata anymore. To make it synced, it selects all fields in metadata at `set_index`.

### CODE REVIEW

File: databricks/koalas/metadata.py [hunk 1]
```diff
@@ -30,69 +32,189 @@ class Metadata(object):
     """
     Manages column names and index information.
 
-    :ivar _column_fields: list of the Spark field names to be seen as columns in Koalas DataFrame.
-    :ivar _index_info: list of pair holding the Spark field names for indexes,
+    :ivar _data_columns: list of the Spark field names to be seen as columns in Koalas DataFrame.
```
Comment (thunterdb) [inline databricks/koalas/metadata.py]: This is super useful doc, thank you for your time writing it. Can you add a comment to say this is an internal class and that it is not meant to be directly accessed by the users (only developers)?

Comment (HyukjinKwon) [reply databricks/koalas/metadata.py]: sure

File: databricks/koalas/metadata.py [hunk 2]
```diff
@@ -30,69 +32,192 @@ class Metadata(object):
...
+
+    Metadata represents the index information for a DataFrame it belongs to. For instance,
+    if we have a Koalas DataFrame as below, Pandas DataFrame does not store the index as columns.
+
+    >>> kdf = ks.DataFrame({
+    ...     'A': [1, 2, 3, 4],
+    ...     'B': [5, 6, 7, 8],
+    ...     'C': [9, 10, 11, 12],
+    ...     'D': [13, 14, 15, 16],
+    ...     'E':[17, 18, 19, 20]})
```
Comment (ueshin) [inline databricks/koalas/metadata.py]: This could be flaky with Python 3.5.

Comment (rxin) [reply databricks/koalas/metadata.py]: also need a space after :

File: databricks/koalas/metadata.py [hunk 3]
```diff
@@ -30,69 +32,192 @@ class Metadata(object):
...
+    >>> kdf1 = kdf.set_index("A")
+    >>> kdf1  # doctest: +NORMALIZE_WHITESPACE
+       B   C   D   E
+    A
+    1  5   9  13  17
+    2  6  10  14  18
+    3  7  11  15  19
+    4  8  12  16  20
+
+    >>> kdf1.to_spark().show()  # doctest: +NORMALIZE_WHITESPACE
```
Comment (rxin) [inline databricks/koalas/metadata.py]: do we just lose the index column when to_spark is called? i think we should keep those. cc @thunterdb

Comment (rxin) [reply databricks/koalas/metadata.py]: oh nvm it's A

**Comment by rxin** [comment]
I like the rest, but I don't think index_pairs is the right name. "pairs" does not describe what that variable is, but rather the current data structure (a pair of objects). It's less intuitive than a name that describes the purpose of the variable. Also if we ever add more information to make it not a "pair", which is pretty likely, we will need to rename a lot of code.

**Comment by HyukjinKwon** [comment]
How about `index_mapping` (since it maps Spark column to Koalas index name?).  I dont mind reverting the naming change too.

**Comment by rxin** [comment]
Yea maybe just index_map.

**Pull request merged**